pkg/agent/clean/active_directory.go

```diff
@@ -0,0 +1,972 @@
+	}
+
+	typemeta := &metav1.ObjectMeta{}
+	err = mapstructure.Decode(metadataMap, typemeta)
+	if err != nil {
+		logrus.Debugf("[%v] errors while decoding typemeta: %v", migrateAdUserOperation, err)
+	}
+
+	storedADConfig.ObjectMeta = *typemeta
+
+	logrus.Debugf("[%v] Should in theory have ActiveDirectory config data? Let's check!", migrateAdUserOperation)
+	logrus.Debugf("[%v] AD Service Account User: %v", migrateAdUserOperation, storedADConfig.ServiceAccountUsername)
+
+	if storedADConfig.ServiceAccountPassword != "" {
+		value, err := common.ReadFromSecret(secrets, storedADConfig.ServiceAccountPassword,
+			strings.ToLower(v3client.ActiveDirectoryConfigFieldServiceAccountPassword))
+		if err != nil {
+			return nil, err
+		}
+		storedADConfig.ServiceAccountPassword = value
+	}
+
+	return storedADConfig, nil
+}
+
+func newCAPool(cert string) (*x509.CertPool, error) {
+	pool, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, err
+	}
+	pool.AppendCertsFromPEM([]byte(cert))
+	return pool, nil
+}
+
+func ldapConnection(config *v3.ActiveDirectoryConfig) (*ldapv3.Conn, error) {
+	caPool, err := newCAPool(config.Certificate)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create caPool: %v", err)
+	}
+
+	servers := config.Servers
+	TLS := config.TLS
+	port := config.Port
+	connectionTimeout := config.ConnectionTimeout
+	startTLS := config.StartTLS
+
+	ldapConn, err := ldap.NewLDAPConn(servers, TLS, startTLS, port, connectionTimeout, caPool)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceAccountUsername := ldap.GetUserExternalID(config.ServiceAccountUsername, config.DefaultLoginDomain)
+	err = ldapConn.Bind(serviceAccountUsername, config.ServiceAccountPassword)
+	if err != nil {
+		return nil, err
+	}
+	return ldapConn, nil
+}
+
+// EscapeUUID will take a UUID string in string form and will add backslashes to every 2nd character.
+// The returned result is the string that needs to be added to the LDAP filter to properly filter
+// by objectGUID, which is stored as binary data.
+func EscapeUUID(s string) string {
+	var buffer bytes.Buffer
+	var n1 = 1
+	var l1 = len(s) - 1
+	buffer.WriteRune('\\')
+	for i, r := range s {
+		buffer.WriteRune(r)
+		if i%2 == n1 && i != l1 {
+			buffer.WriteRune('\\')
+		}
+	}
+	return buffer.String()
+}
+
+func findDistinguishedName(guid string, lConn *ldapv3.Conn, adConfig *v3.ActiveDirectoryConfig) (string, error) {
+	query := fmt.Sprintf("(%v=%v)", "objectGUID", EscapeUUID(guid))
+	search := ldapv3.NewSearchRequest(adConfig.UserSearchBase, ldapv3.ScopeWholeSubtree, ldapv3.NeverDerefAliases,
+		0, 0, false,
+		query, ldap.GetUserSearchAttributes("memberOf", "objectClass", adConfig), nil)
+
+	result, err := lConn.Search(search)
+	if err != nil {
+		return "", err
+	}
+
+	if len(result.Entries) < 1 {
+		return "", LdapErrorNotFound{}
+	} else if len(result.Entries) > 1 {
+		return "", LdapErrorNotFound{}
```

Comment:
GUID is guaranteed to be unique in AD, so I don't think there is a chance of this happening, right?

Reply:
In the unlikely case that this occurs (... is this even active directory? it **might not be**, because delightfully, you can run the AD provider with a non-AD OpenLDAP backend) we should probably consider that a weirder error and maybe bail. It's more likely to indicate a configuration error than anything we can confidently action on.

Reply:
I decided to yell loudly about this, and treat such a user as "skipped" rather than "missing," since I don't know the semantics of upstream AD well enough to know if this might be a possible scenario for a small subset of users. I'd strongly guess that this should never, ever happen. But if it does, we'll know about it. :)